### PR TITLE
Quantile Testing - SQL generation

### PR DESCRIPTION
**Pull request by lukesonnet**
# Features

## Added SQL generation for quantile metrics

This PR adds the ability to generate the necessary quantiles for quantile metrics (unit vs. event, ignoreZeros vs not ignoreZeros) for both optimized and non-optimized fact metric queries. It restructures how we get quantiles to minimize the amount of code we have to reproduce across all the different Integrations.

[This gist](https://gist.github.com/lukesonnet/192b3e3c1ea4b0bfd999ad2f18d38dbf) has code for the 5 kinds of queries:
* 4 single queries for unit w/ and w/o ignore zeros; and event w/ and w/o ignore zeros
* [1 query for a fact optimization](https://gist.github.com/lukesonnet/192b3e3c1ea4b0bfd999ad2f18d38dbf#file-orders-sql) with all 4 in one (and some other metrics you can ignore)

Main changes:

For "event" quantile metrics:
* We build `__quantileMetric` which uses the user metric join, before aggregating, and computes quantiles over every metric row that belongs to our experiment users. We then join these on to the `__userMetricAgg` step  to compute new values in `getAggregateMetricColumn` that we need for our clustered adjustment (we need to compute the N of events per user that are < the quantile AND the N events per user). This is handled in `getAggregateMetricColumn` [[gist example](https://gist.github.com/lukesonnet/192b3e3c1ea4b0bfd999ad2f18d38dbf#file-orders-sql-L238-L245)].
* We also modify `addCaseWhenTimeFilter` to return NULL if the event value is zero when quantile ignorezeros is true [[gist example](https://gist.github.com/lukesonnet/192b3e3c1ea4b0bfd999ad2f18d38dbf#file-orders-sql-L146)].

For "unit" quantile metrics:
* We aggregate at the user level as usual, but now in `__userMetricAgg`, when we call `getAggregateMetricColumn` we do not `COALESCE` for unit + ignoreZeros queries, because we want to return a user level value of `NULL` so that it gets ignored when counting units for `quantile_n` and when computing quantiles. [[gist example](https://gist.github.com/lukesonnet/192b3e3c1ea4b0bfd999ad2f18d38dbf#file-orders-sql-L255)]
* the final quantile computing actually happens in the final select statement.

## Added processing quantiles for grid approach

"grid approach": Following the quantile algorithm we are implementing and only go through the data in one pass on all SQL engines, we need to find a variety of lower and upper bound quantiles and pick the right one after the fact. We use a hard-coded alpha of `0.05` at a variety of values of N (called `N_STAR_VALUES`, and still being finalized) and get the lower and upper bounds as if that was our data size. That means we get `N_STAR_VALUES.length * 2` quantiles for the lower and upper bounds with each query, along with the actual quantile itself.

Because eventually `N_STAR_VALUES` may grow to be size 15 when we release, we are returning 30+ columns per metric, but we only need the 2 columns that correspond to the largest `nstar` such that `nstar < quantile_n`. This happens in the `runExperiment*Query` methods (e.g. https://github.com/growthbook/growthbook/pull/2274/files#diff-929850ea8fdeb86bac06cfd07f3d493e309a318aa3d1514c00299aa5d33d5630R579)

This means the View Queries UI will not show all the columns we compute, which may be confusing, but it also saves DB space because we write less data to mongo, saves bloating our returned data types, and simplifies the python. It also makes it so the View Queries UI shows the most relevant data to our users (if not data that is 1:1 with the SQL output).

TODO:
- [x] ~LIMIT to return 3000 rows~ DID NOT DO, WILL DO IN FOLLOW UP
- [x] Consider forcing all quantiles of a fact table into their own group query to not slow down other optimized fact queries
- [x] Testing below

# Next steps

* Land gbstats changes
* Add REST API, connect to python implementation, release feature flag

# Testing

All testing has only been done for BigQuery, Snowflake, Postgres, Databricks, and ClickHouse.

MySQL, Redshift, Athena, and Presto have not been tested thoroughly.

## Quantile values as expected

* Quantile values are reasonable, but not yet scientifically verified beyond some ad-hoc tests.
* Quantile values are identical across fact and non-fact queries
* Quantile values are the same across engines, modulo some <2% discrepancies in a few edge cases that are to be expected. Discrepancies are very minor within ({postgres, snowflake, clickhouse}) which seem to use similar approaches and ({databricks, bigquery}) which both seem to build an array of quantiles and then sample from that.

## Other results are unchanged (except some capping)

* All other `base` experiment results are unchanged except for some capping. This is to be expected as (1) only the capping logic changed and (2) we started using approximate percentiles for DataBricks and Snowflake while we used exact methods before. Confirmed that outside of those 2 engines, no changes whatsoever.
* `base_pipeline`, one of the conversion window override experiments, and `dimension_user` are all working as expected.

# Screenshots

Properly splits out queries into own group for fact metric optimization and returns right values from processing n vs. nstar:
Screenshot 2024-03-26 at 6 48 04 PM
Screenshot 2024-03-26 at 6 47 59 PM

### CODE REVIEW

File: packages/back-end/src/integrations/SqlIntegration.ts [hunk 1]
```diff
@@ -552,6 +561,32 @@ export default abstract class SqlIntegration
...
+              const upperColumn = `${prefix}quantile_upper_${n}`;
+              if (
+                row[lowerColumn] === undefined ||
+                row[upperColumn] === undefined
+              )
+                return;
+
+              if (
+                n < quantile_n &&
+                n > (Number(metricData[`${prefix}quantile_nstar`]) || 0)
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/back-end/src/integrations/SqlIntegration.ts:580]: Yes, that's right. It's to make sure we don't rely on `N_STAR_VALUES` being in ascending order. I'll add a comment.

File: packages/back-end/src/integrations/SqlIntegration.ts [hunk 2]
```diff
@@ -552,6 +561,32 @@ export default abstract class SqlIntegration
               metricData[prefix + col] = parseFloat(row[prefix + col]) || 0;
             }
           });
+
+          if (row[prefix + "quantile"] !== undefined) {
+            const quantile_n = parseFloat(row[prefix + "quantile_n"]) || 0;
+
+            // process grid for quantile data
+            N_STAR_VALUES.forEach((n) => {
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/back-end/src/integrations/SqlIntegration.ts]: Done. There's an `eslint` ignore line in my new code, so please take a look.

File: packages/back-end/src/queryRunners/ExperimentResultsQueryRunner.ts [hunk 3]
```diff
@@ -75,6 +76,13 @@ export function getFactMetricGroup(metric: FactMetricInterface) {
       return "";
     }
   }
+
+  // Quantile metrics get their own group to prevent slowing down the main query
+  if (quantileMetricType(metric)) {
+    return metric.numerator.factTableId
+      ? `${metric.numerator.factTableId} (quantile metrics)`
+      : "";
+  }
```
Comment (jdorn) [inline packages/back-end/src/queryRunners/ExperimentResultsQueryRunner.ts:85]: Should we only move event quantiles to a separate group?  From what I understand, unit quantiles have no big performance impact on queries, right?

**Review by jdorn — APPROVED** [review]

File: packages/back-end/src/queryRunners/ExperimentResultsQueryRunner.ts (hunk 3 above)
Comment (lukesonnet) [reply packages/back-end/src/queryRunners/ExperimentResultsQueryRunner.ts:85]: They both are similarly problematic as the percentile functions themselves are likely to be expensive.

The event quantile only adds the need to join a small dimensional table (__quantileMetric, one row per variation-dimension) to a large dimensional table (__userMetricJoin one row per user).

I don't think that join should be too bad?
